Public/Src/Engine/Scheduler/Fingerprints/PipFingerprinter.cs

```diff
@@ -260,7 +274,12 @@ protected virtual void AddWeakFingerprint(IFingerprinter fingerprinter, Process
             AddFileOutput(fingerprinter, "StandardError", process.StandardError);
             AddFileOutput(fingerprinter, "StandardOutput", process.StandardOutput);
 
-            fingerprinter.AddOrderIndependentCollection<FileArtifact, ReadOnlyArray<FileArtifact>>("Dependencies", process.Dependencies, (fp, f) => AddFileDependency(fp, f), m_expandedPathFileArtifactComparer);            
+            fingerprinter.AddOrderIndependentCollection<FileArtifact, ReadOnlyArray<FileArtifact>>("Dependencies", process.Dependencies, (fp, f) => {
+                if (!(untrackedExecutable && process.Executable.Path.Equals(f.Path)))    // Don't record executable in weak fingerprint if executable is untracked
+                {
+                    AddFileDependency(fp, f);
+                }
+            }, m_expandedPathFileArtifactComparer);

```

Comment:
I think this is the right way to do it and not mess with ProcessBuilder.  @narasamdya, do you agree?

Reply:
I agree. The only concerning part is complexity, which is quadratic. But how bad it is, I don't know; depend on the size of untrack paths/scopes.

---
In reply to: [332731236](https://github.com/microsoft/BuildXL/pull/997#discussion_r332731236) [](ancestors = 332731236)